python/ray/data/tests/test_streaming_executor.py

```diff
@@ -0,0 +1,173 @@
+    o3 = MapOperator(make_transform(lambda block: [b * 2 for b in block]), o2)
+    topo, _ = build_streaming_topology(o3)
+    assert len(topo) == 3, topo
+    assert o1 in topo, topo
+    assert not topo[o1].inqueues, topo
+    assert topo[o1].outqueue == topo[o2].inqueues[0], topo
+    assert topo[o2].outqueue == topo[o3].inqueues[0], topo
+
+
+def test_process_completed_tasks():
+    inputs = make_ref_bundles([[x] for x in range(20)])
+    o1 = InputDataBuffer(inputs)
+    o2 = MapOperator(make_transform(lambda block: [b * -1 for b in block]), o1)
+    topo, _ = build_streaming_topology(o2)
+
+    # Test processing output bundles.
+    assert len(topo[o1].outqueue) == 0, topo
+    assert process_completed_tasks(topo) is False
+    assert len(topo[o1].outqueue) == 20, topo
+
+    # Test processing completed work items.
+    sleep_ref = sleep.remote()
+    done_ref = ray.put("done")
+    o2.get_work_refs = MagicMock(return_value=[sleep_ref, done_ref])
+    o2.notify_work_completed = MagicMock()
+    o2.inputs_done = MagicMock()
+    assert process_completed_tasks(topo) is True
+    o2.notify_work_completed.assert_called_once_with(done_ref)
+    o2.inputs_done.assert_not_called()
+
+    # Test input finalization.
+    o2.get_work_refs = MagicMock(return_value=[done_ref])
+    o2.notify_work_completed = MagicMock()
+    o2.inputs_done = MagicMock()
+    o1.completed = MagicMock(return_value=True)
+    topo[o1].outqueue.clear()
+    assert process_completed_tasks(topo) is False
+    o2.notify_work_completed.assert_called_once_with(done_ref)
+    o2.inputs_done.assert_called_once_with(input_index=0)
+
+
+def test_select_operator_to_run():
+    inputs = make_ref_bundles([[x] for x in range(20)])
+    o1 = InputDataBuffer(inputs)
+    o2 = MapOperator(make_transform(lambda block: [b * -1 for b in block]), o1)
+    o3 = MapOperator(make_transform(lambda block: [b * 2 for b in block]), o2)
+    topo, _ = build_streaming_topology(o3)
+
+    # Test empty.
+    assert select_operator_to_run(topo) is None
+
+    # Test backpressure based on queue length between operators.
+    topo[o1].outqueue.append("dummy1")
+    assert select_operator_to_run(topo) == o2
+    topo[o1].outqueue.append("dummy2")
+    assert select_operator_to_run(topo) == o2
+    topo[o2].outqueue.append("dummy3")
+    assert select_operator_to_run(topo) == o3
+
+    # Test backpressure includes num active tasks as well.
+    topo[o3].num_active_tasks = MagicMock(return_value=2)
+    assert select_operator_to_run(topo) == o2
+    topo[o2].num_active_tasks = MagicMock(return_value=2)
+    assert select_operator_to_run(topo) == o3
+
+
+def test_dispatch_next_task():
+    inputs = make_ref_bundles([[x] for x in range(20)])
+    o1 = InputDataBuffer(inputs)
+    o2 = MapOperator(make_transform(lambda block: [b * -1 for b in block]), o1)
+    op_state = OpState(o2)
+    o2.add_input = MagicMock()
+    op_state.inqueues[0].append("dummy1")
+    dispatch_next_task(op_state)
```

Comment:
Suggest running this multiple times so we can test indices other than 0.

Reply:
Added a test for multiple inputs, and a TODO for multiple indices.